### PR TITLE
[exporter/elasticsearch] Replace go-elasticsearch BulkIndexer with go-docappender

### LINKED ISSUE #32378: [exporter/elasticsearch] Replace go-elasticsearch with go-docappender BulkIndexer to prepare for reliability fixes
### Component(s)

exporter/elasticsearch

### Is your feature request related to a problem? Please describe.

This is required by https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32377 to fix https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30792.

go-elasticsearch BulkIndexer does not have the Flush function required by exporterhelper batch sender to guarantee that a bulk index request is sent.

### Describe the solution you'd like

Replace go-elasticsearch with go-docappender for the Flush function

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by carsonip**
**Description:** <Describe what has changed.>

Replace go-elasticsearch BulkIndexer with go-docappender BulkIndexer for Flush function in preparation for reliability fixes. Maintain similar interface and implementation to go-elasticsearch BulkIndexer.

Further changes to expose individual `docappender.BulkIndexer` instances are needed down the road but it is out of the scope of this PR.

Implications of this change:
- flush timeout is now enforced on client side
- oversize payload special handling is now removed
- go-docappender uses bulk request filterPath which means bulk response is smaller, less JSON parsing and lower CPU usage
- document level retry debug logging is removed as retries are done transparently

~~Blocked by #32585~~

**Link to tracking Issue:** <Issue number if applicable> Fixes #32378 

**Testing:** Integration test is passing

### CODE REVIEW

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (carsonip) [inline exporter/elasticsearchexporter/elasticsearch_bulk.go]: [To reviewer] This is a change in default as well as the performance under default config.

NumCPU was the original default used by go-elasticsearch. But I think that setting the default to 1 makes more sense and 1 is also the default in opensearch exporter (see [code](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/opensearchexporter/log_bulk_indexer.go#L120)).

File: exporter/elasticsearchexporter/go.mod
Comment (carsonip) [inline exporter/elasticsearchexporter/go.mod:99]: Blocked by https://github.com/elastic/go-docappender/pull/146

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (carsonip) [reply exporter/elasticsearchexporter/elasticsearch_bulk.go:146]: Blocked by https://github.com/elastic/go-docappender/pull/155

File: exporter/elasticsearchexporter/go.mod
Comment (carsonip) [reply exporter/elasticsearchexporter/go.mod]: Done

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (carsonip) [reply exporter/elasticsearchexporter/elasticsearch_bulk.go]: Will change the default in a follow up PR

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (lahsivjar) [inline exporter/elasticsearchexporter/elasticsearch_bulk.go:217]: What do you think about designing the `BulkIndexerPool` with a channel containing `NumWorkers` count of bulk indexers? For adding we can take a bulk indexer from the channel and add stuff to it. Once done we can decide to flush and put it back in the pool. Another option for flush could be to just flush the indexers one by one periodically in a single goroutine defined on the `BulkIndexerPool` itself.

Comment (carsonip) [reply exporter/elasticsearchexporter/elasticsearch_bulk.go:217]: I have thought about this. While it'll be nice to have this enhancement, the current BulkIndexerPool is the closest to the go-elasticsearch BulkIndexer implementation, so it will make this PR easier to review. Also, with https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32377 the flush behavior and mechanism will be vastly different. I'm not sure if it is worth it to fix the sparse bulk request behavior now.

File: exporter/elasticsearchexporter/README.md
Comment (carsonip) [inline exporter/elasticsearchexporter/README.md:54]: [to reviewer] There is no change in practice and the README was wrong. The number came from https://github.com/elastic/go-elasticsearch/blob/7.17/esutil/bulk_indexer.go#L204

**Comment by lahsivjar** [comment]
There is a bug in the go-docappender causing the elasticsearch exporter's integration test to fail. I have opened a PR with the fix: https://github.com/elastic/go-docappender/pull/160

**Comment by carsonip** [comment]
Waiting for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32585 to merge, then it should come out of draft.

File: exporter/elasticsearchexporter/logs_exporter.go
Comment (ycombinator) [inline exporter/elasticsearchexporter/logs_exporter.go:73]: Aren't these breaking changes for the user?

Comment (carsonip) [reply exporter/elasticsearchexporter/logs_exporter.go:73]: No, there is no functional change here. It is no longer storing maxAttempts and retryOnStatus separately within the exporter because they are already passed into elasticsearch client (for request level retries) and bulk indexer (for document level retries) in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32359/files#diff-a09f63572f3eb4cbff2f4ef096a2cd5c71025e6e3c882617d466f57120aeefd8R36

Comment (ycombinator) [reply exporter/elasticsearchexporter/logs_exporter.go:73]: Yep, I see it now.  Thanks!

**Review by ycombinator — APPROVED** [review]
LGTM but please wait for @lahsivjar's LGTM as well.  Thanks!

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (lahsivjar) [inline exporter/elasticsearchexporter/elasticsearch_bulk.go:282]:
```suggestion
	flushTick := time.NewTicker(w.flushInterval)
	defer flushTick.Stop()
```

**Review by lahsivjar — APPROVED** [review]
Looks good overall, just some minor comments.

File: exporter/elasticsearchexporter/elasticsearch_bulk.go
Comment (carsonip) [reply exporter/elasticsearchexporter/elasticsearch_bulk.go:203]: - Used 0 to avoid an import, but I agree that gzip.NoCompression will be clearer

> Also, I see that this field is not exposed as a config, what is the reasoning behind going with no compression?

I do not want to introduce any behavioral change on compression in this PR. It can be done in a follow up PR to expose compression config.

Comment (lahsivjar) [reply exporter/elasticsearchexporter/elasticsearch_bulk.go]: > Used 0 to avoid an import, especially when docappender does not use the stock gzip

We can just not set the `CompressionLevel` in that case (a lot clearer IMO)

Also, docappender uses “github.com/klauspost/compress/gzip”` import but it should already be in the indirect dependency tree.

**Review by andrzej-stencel — APPROVED** [review]

**Comment by BenEfrati** [comment]
Hi @carsonip , @andrzej-stencel 

There is a regression after this change - duplicate key in json:
before this change request to ELK was:
```
{
...
   "host":{
      "name":"hostname"
   },
   "os":{
      "description":"Windows 11 10.0",
      "type":"windows"
   },
...
}
```

after this change:
```
{
...
   "host":{
      "hostname":"hostname",
      "os":{
         "full":"Windows 11 10.0",
         "platform":"windows"
      }
   },
...
 "host":{
      "os":{
         "type":"windows"
      }
   },
...
}
```

in this example `host` key is duplicated, result in invalid json and error from ELK
```
 "error": {
  "reason": "failed to parse",
  "type": "mapper_parsing_exception"
  },
  "status": 400
```

**Pull request merged**
